src/Microsoft.Azure.WebJobs.Host/Blobs/Listeners/ScanBlobScanLogHybridPollingStrategy.cs

```diff
@@ -0,0 +1,162 @@
+            ScnaContainersAndRegisterWithStrategy(_cancellationTokenSource.Token);
+            _scanContainerStrategy.Start();
+            _pollLogStrategy.Start();
+        }
+
+        public void Cancel()
+        {
+            ThrowIfDisposed();
+            _scanContainerStrategy.Cancel();
+            _pollLogStrategy.Cancel();
+            _cancellationTokenSource.Cancel();
+        }
+
+        public Task RegisterAsync(IStorageBlobContainer container, ITriggerExecutor<IStorageBlob> triggerExecutor, CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+            // Register and Execute are not concurrency-safe.
+            // Avoiding calling Register while Execute is running is the caller's responsibility.
+
+            ICollection<ITriggerExecutor<IStorageBlob>> containerRegistrations;
+            if (_registrations.ContainsKey(container))
+            {
+                containerRegistrations = _registrations[container];
+            }
+            else
+            {
+                containerRegistrations = new List<ITriggerExecutor<IStorageBlob>>();
+                _registrations.Add(container, containerRegistrations);
+            }
+
+            containerRegistrations.Add(triggerExecutor);
+            return Task.FromResult(0);
+        }
+
+        public async Task<TaskSeriesCommandResult> ExecuteAsync(CancellationToken cancellationToken)
+        {
+            ThrowIfDisposed();
+            await Task.WhenAll(_scanContainerStrategy.ExecuteAsync(cancellationToken),
+                                _pollLogStrategy.ExecuteAsync(cancellationToken));
+
+            // Run subsequent iterations at 2 second intervals.
+            return new TaskSeriesCommandResult(wait: Task.Delay(TwoSeconds));
+        }
+
+        public void Notify(IStorageBlob blobWritten)
+        {
+            ThrowIfDisposed();
+            _scanContainerStrategy.Notify(blobWritten);
+            _pollLogStrategy.Notify(blobWritten);
+        }
+
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                _pollLogStrategy.Dispose();
+                _scanContainerStrategy.Dispose();
+                _cancellationTokenSource.Dispose();
+                _disposed = true;
+            }
+        }
+
+        private void ScnaContainersAndRegisterWithStrategy(CancellationToken cancellationToken)
+        {
+            foreach (IStorageBlobContainer container in _registrations.Keys)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return;
+                }
+
+                IStorageBlobResultSegment results;
+
+                try
+                {
+                    // Non-async is correct here. ScanContainers occurs on a background thread. Unless it blocks, no one
+                    // else is around to observe the results.
```

Comment:
+1.  I'm assuming the problem is that Start(), which calls this, is synchronous and needs to block. If so, can you make this method async / await and push the blocking up up the stack? 

---
In reply to: [54306524](https://github.com/Azure/azure-webjobs-sdk/pull/637#discussion_r54306524) [](ancestors = 54306524)